src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java

```diff
@@ -0,0 +1,159 @@
+        return index.list(packageFile);
+    }
+
+
+    private void indexURLs(URL url) {
+       try {
+           if (!"file".equals(url.getProtocol())) {
+               throw new RuntimeException("Classpath element is not a file: " + url);
+           }
+           File root = new File(url.getPath());
+           URL[] args = {url};
+           URLClassPath delegate = new URLClassPath(args);
+
+           if (root.isDirectory()) {
+               String rootPath = root.getPath();
+               if (!rootPath.endsWith(File.separator)) {
+                   rootPath = rootPath + File.separator;
+               }
+               addFilesToIndex(rootPath.length(), root, delegate);
+           } else if (root.isFile() && root.getName().endsWith(".jar")) {
+               JarFile jarFile = new JarFile(root);
+               try {
+                   Enumeration<JarEntry> entries = jarFile.entries();
+                   while (entries.hasMoreElements()) {
+                       JarEntry entry = entries.nextElement();
+                       String name = entry.getName();
+                       maybeIndexResource(name, delegate);
+                   }
+               } finally {
+                   jarFile.close();
+               }
+           }
+       } catch (IOException e) {
+           throw new RuntimeException(e);
+       }
+    }
+
+    private void addFilesToIndex(int basePrefixLen, File f, URLClassPath delegate) throws IOException {
+        if (f.isDirectory()) {
+            if (f.getPath().length() > basePrefixLen) {  // Don't index the root itself.
+                String relPath = f.getPath().substring(basePrefixLen);
+                maybeIndexResource(relPath, delegate);
+            }
+            File[] directoryEntries = f.listFiles();
+            assert(directoryEntries != null);
+            for (int i = 0; i < directoryEntries.length; ++i) {
+                addFilesToIndex(basePrefixLen, directoryEntries[i], delegate);
+            }
+        } else {
+            String relPath = f.getPath().substring(basePrefixLen);
+            maybeIndexResource(relPath, delegate);
+        }
+    }
+
+    private void maybeIndexResource(String relPath, URLClassPath delegate) {
+
+        if (!index.containsKey(relPath)) {
+            index.put(relPath, delegate);
+            // Callers may request the directory itself as a resource, and may
+            // do so with or without trailing slashes.  We do this in a while-loop
+            // in case the classpath element has multiple superfluous trailing slashes.
+            if (relPath.endsWith(File.separator)) {
+                maybeIndexResource(relPath.substring(0, relPath.length() - File.separator.length()), delegate);
+            }
+        }
+    }
+
+
+    // Map from resource name to URLClassPath to delegate loading that resource to.
+    private final ClassPathTree index = new ClassPathTree("/", "/", new HashMap<String, ClassPathTree>(), null);
```

Comment:
I expected this declaration at beginning of class.

Reply:
f385baae40686db0f1cf6db42cf9730ec4aec7ab